File: java/src/main/java/io/github/matheuslm7/brvalidator/ie/Ms.java
package io.github.matheuslm7.brvalidator.ie;

import io.github.matheuslm7.brvalidator.internal.Shared;
import java.util.regex.Pattern;

/**
 * MS: verified against the official SEFAZ-MS "Roteiro de Crítica da Inscrição Estadual"
 * (sintegra.gov.br/Cad_Estados/cad_MS.html) for the algorithm (which the page does not
 * accompany with a worked numeric example), plus an independently hand-computed regression
 * vector (281234566). Format: 8 digits (always starting with "28" or "50") + 1 check digit.
 */
final class Ms {

  private static final Pattern ALLOWED_CHARS = Pattern.compile("^[\\d\\s]+$");
  private static final int[] WEIGHTS = {9, 8, 7, 6, 5, 4, 3, 2};

  private Ms() {}

  static String normalize(String value) {
    return Shared.removeNonDigits(value);
  }

  private static boolean hasValidPrefix(String v) {
    String prefix = v.substring(0, 2);
    return prefix.equals("28") || prefix.equals("50");
  }

  static boolean isValid(String value) {
    if (!ALLOWED_CHARS.matcher(value).matches()) {
      return false;
    }

    String v = normalize(value);

    if (v.length() != 9 || !hasValidPrefix(v)) {
      return false;
    }

    return Mod11.mod11CheckDigit(Mod11.weightedSum(v, WEIGHTS)) == (v.charAt(8) - '0');
  }

  static String format(String value) {
    return normalize(value);
  }
}

File: src/ie/states/ms.ts
import { removeNonDigits } from "../../shared/normalize.js";
import { weightedSum, mod11CheckDigit } from "../../shared/mod11.js";

// Verified against the official SEFAZ-MS "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_MS.html) for the algorithm
// (which the page does not accompany with a worked numeric example), plus
// an independently hand-computed regression vector (281234566). Format: 8
// digits (always starting with "28" or "50") + 1 check digit.
const WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2];

function normalize(value: string): string {
  return removeNonDigits(value);
}

function hasValidPrefix(ie: string): boolean {
  const prefix = ie.slice(0, 2);

  return prefix === "28" || prefix === "50";
}

function isValid(value: string): boolean {
  if (!/^[\d\s]+$/.test(value)) {
    return false;
  }

  const ie = normalize(value);

  if (ie.length !== 9 || !hasValidPrefix(ie)) {
    return false;
  }

  return mod11CheckDigit(weightedSum(ie, WEIGHTS)) === Number(ie[8]);
}

function format(value: string): string {
  return normalize(value);
}

export const MS = {
  isValid,
  normalize,
  format,
};
